src/zenml/steps/step_context.py

```diff
@@ -25,57 +25,73 @@
 )
 
 from zenml.client import Client
-from zenml.environment import Environment
 from zenml.exceptions import StepContextError
+from zenml.logger import get_logger
+from zenml.utils.singleton import SingletonMetaClass
 
 if TYPE_CHECKING:
     from zenml.config.step_run_info import StepRunInfo
     from zenml.materializers.base_materializer import BaseMaterializer
-    from zenml.stack import Stack
 
+logger = get_logger(__name__)
 
-class StepContextOutput(NamedTuple):
-    """Tuple containing materializer class and URI for a step output."""
 
-    materializer_classes: Sequence[Type["BaseMaterializer"]]
-    artifact_uri: str
+def get_step_context() -> "StepContext":
+    """Get the context of the currently running step.
 
+    Returns:
+        The context of the currently running step.
 
-class StepContext:
+    Raises:
+        RuntimeError: If no step is currently running.
+    """
+    if not isinstance(StepContext.__singleton_instance, StepContext):
+        raise RuntimeError(
+            "The step context is only available inside a step function."
+        )
+    return StepContext.__singleton_instance
+
+
+class StepContext(metaclass=SingletonMetaClass):
```

Comment:
I think the `StepContext` could offer some more convenience properties/methods to make access to some things easier?

Reply:
- All steps of the current pipeline run (I guess that is solved on the Post-execution branch already?)
- Immediate upstream step runs
- Probably other things that are annoying to fetch by querying the models (but I'm not sure how this looks with the updated models):
```python
run = get_step_context().pipeline_run
if run.pipeline:
  pipeline_name = run.pipeline.name
else:
  ...
```

Reply:
IMO i would rather implement those convencience properties on the models directly, then you can use it in post execution and other places as well

Reply:
Other steps of the run is already solved via `pipeline_run.steps` on post exec, yes

Reply:
Immediate upstream steps is not easy to get right now. We have parent step IDs in the step models so I could add another `step.parent_steps` property similar to the other post exec changes if you think that's useful

Reply:
Added that property on #1635